pysilcam/tracking/silcamtracker.py

```diff
@@ -0,0 +1,422 @@
+                # calculation
+                cr2 = props2[int(idx - 1)].centroid  # subtract 1 from idx because of zero-indexing
+
+                # remove this particle from future calculations
+                iml2[iml2 == int(idx)] = 0
+
+                x.append(cr2[1])  # col
+                y.append(cr2[0])  # row
+
+                # and append the position of this particle from the image 1
+                y1.append(cr[0])  # row
+                x1.append(cr[1])  # col
+
+                # if we get here then we also want the particle stats
+                ecd.append(el.equivalent_diameter)
+                length.append(el.major_axis_length)
+                width.append(el.minor_axis_length)
+                break
+
+    X = [x1, x]  # horizontal vector
+    Y = [y1, y]  # vertical vector
+    return X, Y, ecd, length, width, imbw_out
+
+
+def get_idx_from_search_box(ecd_tolerance, ecd_lookup, search_roi):
+    '''
+
+    :param ecd_tolerance:
+    :param ecd_lookup:
+    :param search_roi:
+    :return:
+    '''
+    # we are still looking for a particle
+    particle_found = False
+    idx = 0
+
+    # and squash to an array of particle indicies from within the box
+    unp = np.unique(search_roi)
+    # size the particles in the search_roi
+    if len(unp) > 1:  # len(unp) == 1 implies no particles
+        props = regionprops(search_roi, cache=True)
+
+        # list all the choices of particle size within the search area
+        choice_ecd = np.array([p.equivalent_diameter for p in props])
+
+        # find the closest ecd
+        closest_ecd = np.min(np.abs(choice_ecd - ecd_lookup))
+        closest_ecd_pcent = closest_ecd / ecd_lookup * 100
+
+        if (closest_ecd_pcent < ecd_tolerance):
+            idx = (np.abs(choice_ecd - ecd_lookup)).argmin()  # find the ecd that is closest
+
+            # get the labelled particle number at this location
+            idx = int(unp[int(idx + 1)])
+            particle_found = True
+    return idx, particle_found
+
+
+def find_particle_idx(cross_correlation_threshold, roi, search_roi, ecd_tolerance, ecd_lookup):
+    '''
+
+    :param cross_correlation_threshold:
+    :param iml2:
+    :param roi:
+    :param search_roi:
+    :param ecd_tolerance:
+    :param ecd_lookup:
+    :return:
+    '''
+    # use a Fast Normalized Cross-Correlation
+    cross_correlation = match_template(search_roi > 0, roi)
+
+    idx = 0
+    particle_found = False
+    if np.max(cross_correlation) > cross_correlation_threshold:
+        particle_found = True
+
+        # look for a peak in the cross-correlation regardless of how good it is
+        # and extract the location of the maximum
+        ij = np.unravel_index(np.argmax(cross_correlation), cross_correlation.shape)
+        x_, y_ = ij[::-1]
+        idx = int(search_roi[int(y_), int(x_)])
+
+        # convert position from inside the bounding box to a position within
+        # the original image
+        #x_ += ((bbox[3] - bbox[1]) / 2)

```

Comment:
E265: block comment should start with '# '  <!-- Automatically posted by Lintly -->